File: 0643-maximum-average-subarray-i/0643-maximum-average-subarray-i.cpp
#include <algorithm>
#include <vector>

class Solution {
public:
    double findMaxAverage(std::vector<int>& nums, int k) {
        // Initialize the sum for the first k elements
        double sum = 0;
        for(int i = 0; i < k; i++) {
            sum += nums[i];
        }

        // Store the initial sum as the maximum average
        double max_avg = sum / k;

        // Slide the window over the array
        for(int i = k; i < nums.size(); i++) {
            // Add the current element and subtract the first element of the previous window
            sum += nums[i] - nums[i - k];
            // Update the maximum average if the new average is greater
            max_avg = std::max(max_avg, sum / k);
        }

        return max_avg;
    }
};

File: 0643-maximum-average-subarray-i/0643-maximum-average-subarray-i.py
# SOL1 - GPT4

class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        # Initialize the sum for the first k elements
        sum_k = sum(nums[:k])
        
        # Store the initial sum as the maximum average
        max_avg = sum_k / k

        # Slide the window over the array
        for i in range(k, len(nums)):
            # Add the current element and subtract the first element of the previous window
            sum_k += nums[i] - nums[i - k]
            # Update the maximum average if the new average is greater
            max_avg = max(max_avg, sum_k / k)

        return max_avg
